File: gyunseo/BOJ/2178/2178.cpp
#include <iostream>
#include <queue>
#include <tuple>
#define endl '\n'
using namespace std;
const int MAX = 100;
queue<pair<int, int>> q;
string board[MAX];
// dist가 is_visited의 역할도 같이한다.
int dist[MAX][MAX];
int di[4] = {-1, 1, 0, 0};
int dj[4] = {0, 0, -1, 1};
int N, M;

bool OOB(int i, int j) {
  if (i < 0 || i >= N) return true;
  if (j < 0 || j >= M) return true;
  return false;
}

void ReadUserInput() {
  cin >> N >> M;
  for (int i = 0; i < N; i++) {
    cin >> board[i];
  }
}

void Solve() {
  int cur_i, cur_j;
  q.push({0, 0});
  dist[0][0] = 1;
  while(!q.empty()) {
    tie(cur_i, cur_j) = q.front();
    q.pop();
    if (cur_i == (N - 1) && cur_j == (M - 1)) {
      cout << dist[cur_i][cur_j] << endl;
      return;
    }
    for (int k = 0; k < 4; k++) {
      int next_i, next_j;
      next_i = cur_i + di[k];
      next_j = cur_j + dj[k];
      if (OOB(next_i, next_j)) continue;
      if (board[next_i][next_j] == '0') continue;
      if (dist[next_i][next_j] >= 1) continue;
      dist[next_i][next_j] = dist[cur_i][cur_j] + 1;
      q.push({next_i, next_j});
    }
  }
}
int main() {
  ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
  ReadUserInput();
  Solve();
  return 0;
}

File: gyunseo/BOJ/2178/2178.py
import sys
from collections import deque

input = sys.stdin.readline
print = sys.stdout.write
MAX_NUM = 100
N, M = map(int, input().rstrip().split())
board = []
is_visited = [[False for _ in range(MAX_NUM)] for __ in range(MAX_NUM)]
di = [-1, 1, 0, 0]
dj = [0, 0, -1, 1]
for _ in range(N):
    board.append(list(map(int, input().rstrip())))


def OOB(i, j):
    if i < 0 or i >= N:
        return True
    if j < 0 or j >= M:
        return True
    return False


q = deque()
q.append((0, 0, 1))
while q:
    cur_i, cur_j, cur_cnt = q.popleft()
    if is_visited[cur_i][cur_j]:
        continue
    is_visited[cur_i][cur_j] = True
    if cur_i == N - 1 and cur_j == M - 1:
        print(f"{cur_cnt}\n")
        break
    for k in range(4):
        next_i, next_j = cur_i + di[k], cur_j + dj[k]
        if OOB(next_i, next_j):
            continue
        if is_visited[next_i][next_j]:
            continue
        if board[next_i][next_j] == 0:
            continue
        q.append((next_i, next_j, cur_cnt + 1))
